### PR TITLE
[risk=low][RW-5638] use MapStruct in getUserRecentResources()

### PR DESCRIPTION
**Pull request by jmthibault79**
Description:

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java [hunk 1]
```diff
@@ -34,8 +37,7 @@
   @Mapping(target = "cdrVersionId", source = "dbWorkspace.cdrVersion")
   @Mapping(target = "accessTierShortName", source = "dbWorkspace.cdrVersion.accessTier.shortName")
   @Mapping(target = "permission", source = "accessLevel")
-  WorkspaceFields workspaceResourceFromWorkspace(
-      DbWorkspace dbWorkspace, WorkspaceAccessLevel accessLevel);
+  WorkspaceFields fromWorkspace(DbWorkspace dbWorkspace, WorkspaceAccessLevel accessLevel);
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java:40]: alt: these could  be `toModel()`

File: api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java [hunk 2]
```diff
@@ -45,60 +47,85 @@ WorkspaceFields workspaceResourceFromWorkspace(
...
+
+  default WorkspaceResource fromDbUserRecentResource(
+      DbUserRecentResource dbUserRecentResource,
+      WorkspaceAccessLevel accessLevel,
+      DbWorkspace dbWorkspace) {
+    return mergeWorkspaceAndResourceFields(
+        fromWorkspace(dbWorkspace, accessLevel), fromDbUserRecentResource(dbUserRecentResource));
+  }
+
+  default FileDetail convertStringToFileDetail(String str) {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java:118]: moved this from UserMetricsController

File: api/src/test/java/org/pmiops/workbench/api/UserMetricsControllerTest.java [hunk 3]
```diff
@@ -60,13 +68,15 @@
   @Mock private WorkspaceDao workspaceDao;
   @Mock private WorkspaceAuthService workspaceAuthService;
 
-  private UserMetricsController userMetricsController;
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/api/UserMetricsControllerTest.java:63]: moved this to bottom, alphabetized the others

File: api/src/test/java/org/pmiops/workbench/api/UserMetricsControllerTest.java [hunk 4]
```diff
@@ -128,17 +146,12 @@ public void setUp() {
     dbUserRecentResource1.setUserId(dbUser.getUserId());
     dbUserRecentResource1.setWorkspaceId(dbWorkspace1.getWorkspaceId());
 
-    when(workspaceDao.findActiveByWorkspaceId(dbUserRecentResource1.getWorkspaceId()))
```
Comment (jmthibault79) [inline api/src/test/java/org/pmiops/workbench/api/UserMetricsControllerTest.java:131]: consolidated these as `mockResponsesForWorkspace()`

File: api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceFields.java [hunk 5]
```diff
@@ -13,7 +13,6 @@
   private BillingStatus workspaceBillingStatus;
   private String cdrVersionId;
   private String accessTierShortName;
-  private String permission;
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceFields.java:16]: this field isn't derived from `DbWorkspace`, and it became awkward to include it here

File: api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java [hunk 6]
```diff
@@ -45,60 +48,96 @@ WorkspaceFields workspaceResourceFromWorkspace(
...
+      source = "fcWorkspace.accessLevel",
+      qualifiedByName = "fcToApiWorkspaceAccessLevel")
   WorkspaceResource mergeWorkspaceAndResourceFields(
-      WorkspaceFields workspaceFields, ResourceFields resourceFields);
+      WorkspaceFields workspaceFields,
+      FirecloudWorkspaceResponse fcWorkspace,
+      ResourceFields resourceFields);
 
-  default WorkspaceResource dbWorkspaceAndDbCohortToWorkspaceResource(
+  default WorkspaceResource fromDbCohort(
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java:99]: could rename these `toModel()` or `dbCohortToModel()` etc

File: api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java [hunk 7]
```diff
@@ -45,60 +48,96 @@ WorkspaceFields workspaceResourceFromWorkspace(
...
+  }
+
+  default FileDetail convertStringToFileDetail(String str) {
+    if (str == null) {
+      return null;
+    }
+    if (!str.startsWith("gs://")) {
+      return null;
+    }
+    int filenameStart = str.lastIndexOf('/') + 1;
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java:138]: Yes this is the new logic from #5944

File: api/src/main/java/org/pmiops/workbench/utils/mappers/FirecloudMapper.java [hunk 8]
```diff
@@ -12,6 +13,7 @@ default WorkspaceAccessLevel fcAccessLevelToApiAccessLevel(FirecloudWorkspaceAcc
     return WorkspaceAccessLevel.fromValue(acl.getAccessLevel());
   }
 
+  @Named("fcToApiWorkspaceAccessLevel")
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/utils/mappers/FirecloudMapper.java:16]: so WorkspaceResourceMapper can refer to it

File: api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java [hunk 9]
```diff
@@ -13,124 +13,63 @@
...
 import org.pmiops.workbench.workspaces.WorkspaceAuthService;
+import org.pmiops.workbench.workspaces.resources.WorkspaceResourceMapper;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
 public class UserMetricsController implements UserMetricsApiDelegate {
 
-  private static final Logger logger = Logger.getLogger(UserMetricsController.class.getName());
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java:49]: heh, not sure why we had two loggers defined

File: api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java [hunk 10]
```diff
@@ -13,124 +13,63 @@
...
-
   private static final int MAX_RECENT_NOTEBOOKS = 8;
+
   private static final Logger log = Logger.getLogger(UserMetricsController.class.getName());
+
+  private final CloudStorageClient cloudStorageClient;
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java:47]: updated and alphabetized

File: api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java [hunk 11]
```diff
@@ -13,124 +13,63 @@
...
   private final UserRecentResourceService userRecentResourceService;
-  private final WorkspaceDao workspaceDao;
   private final WorkspaceAuthService workspaceAuthService;
-  private final FireCloudService fireCloudService;
-  private final CloudStorageClient cloudStorageClient;
-  private final CommonMappers commonMappers;
-  private FirecloudMapper firecloudMapper;
-  private int distinctWorkspaceLimit = 5;
-
-  // TODO(jaycarlton): migrate these private functions to MapStruct
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java:63]: we've been wanting to do this for a while

File: api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java [hunk 12]
```diff
@@ -13,124 +13,63 @@
...
-  // Converts DB model to client Model
-  // RW-5638
-  private final Function<DbUserRecentResource, WorkspaceResource> TO_CLIENT =
-      userRecentResource -> {
-        WorkspaceResource resource = new WorkspaceResource();
-        resource.setCohort(TO_CLIENT_COHORT.apply(userRecentResource.getCohort()));
-        resource.setConceptSet(TO_CLIENT_CONCEPT_SET.apply(userRecentResource.getConceptSet()));
-        FileDetail fileDetail = convertStringToFileDetail(userRecentResource.getNotebookName());
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java:71]: this is the only line that's not a straightforward assignment or conversion.  This is now in WorkspaceResourceMapper.

File: api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java [hunk 13]
```diff
@@ -143,25 +82,28 @@ public void setDistinctWorkspaceLimit(int limit) {
       String workspaceNamespace, String workspaceId, RecentResourceRequest recentResourceRequest) {
     workspaceAuthService.enforceWorkspaceAccessLevel(
         workspaceNamespace, workspaceId, WorkspaceAccessLevel.WRITER);
-    // this is only ever used for Notebooks because we update/add to the cache for the other
-    // resources in the backend
-    // Because we don't store notebooks in our database the way we do other resources.
-    long wId = getWorkspaceId(workspaceNamespace, workspaceId);
-    String notebookPath;
+
+    final FirecloudWorkspaceResponse fcWorkspace =
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java:86]: we always need the fcWorkspace now, so I moved it up out of the conditional

File: api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java [hunk 14]
```diff
@@ -143,25 +82,28 @@ public void setDistinctWorkspaceLimit(int limit) {
...
-          fireCloudService
-              .getWorkspace(workspaceNamespace, workspaceId)
-              .getWorkspace()
-              .getBucketName();
+      String bucket = fcWorkspace.getWorkspace().getBucketName();
       notebookPath = "gs://" + bucket + "/notebooks/" + recentResourceRequest.getNotebookName();
     }
-    DbUserRecentResource recentResource =
+
+    // this is only ever used for Notebooks because we update/add to the cache for the other
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java:97]: not sure if I moved this comment to the right place

File: api/src/main/java/org/pmiops/workbench/utils/mappers/WorkspaceMapper.java [hunk 15]
```diff
@@ -54,6 +54,10 @@
   @Mapping(target = "accessTierShortName", source = "dbWorkspace.cdrVersion.accessTier.shortName")
   Workspace toApiWorkspace(DbWorkspace dbWorkspace);
 
+  @Mapping(
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/utils/mappers/WorkspaceMapper.java:57]: not sure why this became necessary, but MapStruct stopped converting plain strings.

`WorkspaceMapperTest.testConvertsFirecloudResponseToApiResponse()` saved us by detecting this

File: api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java [hunk 16]
```diff
@@ -45,60 +48,96 @@ WorkspaceFields workspaceResourceFromWorkspace(
...
   @Mapping(target = "cohort", ignore = true)
   @Mapping(target = "cohortReview", ignore = true)
   @Mapping(target = "conceptSet", ignore = true)
   @Mapping(target = "notebook", ignore = true)
   @Mapping(target = "dataSet", source = "dbDataset", qualifiedByName = "dbModelToClientLight")
   @Mapping(target = "lastModifiedEpochMillis", source = "dbDataset.lastModifiedTime")
-  ResourceFields workspaceResourceFromDbDataset(DbDataset dbDataset);
+  ResourceFields fromDbDataset(DbDataset dbDataset);
+
+  // TODO why are Cohort Review and Dataset not present in DbUserRecentResource?
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java:77]: Discussion at https://pmi-engteam.slack.com/archives/CHRN2R51N/p1638225087021900

**Review by yonghaoy — APPROVED** [review]

**Pull request merged**